### ISSUE TITLE
Explore using the interactive window as a console for notebook kernels

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/vscode-jupyter#6765]
From discussion with @greazer @rchiodo @claudiaregio @rebornix

Existing feature requests: 
- https://github.com/microsoft/vscode-jupyter/issues/4573
    - "I add some cells to try something out and I don't want to lose them, but I don't want them in my 'main' notebook."
- https://github.com/microsoft/vscode-jupyter/issues/6484 
    - "...interactive command line that is connected to a jupyter notebook..."
- https://github.com/microsoft/vscode-jupyter/issues/6033
    - "...create interactive Python consoles connected to specific notebook kernels. The main usecase is to be able to work interactively based on the state of a notebook without having to tinker with the notebook itself."
    - "...have an ability to launch an IPython shell (within VS Code) which is binded to the kernel of a specific notebook (so all objects defined in the notebook are accessible there etc.)..."

Proposed experience:
- Command contributed to `notebook/cell/title` to open interactive window connected to same kernel
- Command contributed to `interactive/cell/title` to paste code back into notebook

Bare-bones prototype:
![recording (7)](https://user-images.githubusercontent.com/30305945/126676422-1bb1beb0-e92c-41eb-97ee-0455b122e40d.gif)

Open questions
- Should users be able to start multiple interactive consoles? Original prototype has scratchpad singleton
- Should the scratchpad contain just the one cell (fully matching https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/scratchpad/README.html behavior) or be a full interactive window?
- Should the interactive window immediately receive focus when created?
- Introduce keybinding or similar to copy code (otherwise user has to move mouse literally the entire span of the screen)

Engineering notes:
- Summary: Jupyter extension needs to refactor our `VSCodeNotebookController` wrapper, `IKernel`, and the cell execution codepath to allow
    - an `IKernel` that is currently mapped to a `jupyter-notebook` controller which has been selected for a given `jupyter-notebook` NotebookDocument 
    - to be reused for an `interactive` controller that is selected for a given `interactive` NotebookDocument
- Current state/background
    - Currently vscode.NotebookControllers are per-notebookType, so to get our kernels to show up for interactive windows, each of our kernels is registered twice with VS Code--once for the `jupyter-notebook` notebookType and once for the `interactive` notebookType
    - Our code currently assumes a 1-1 relationship between a VS Code NotebookController and our internal IKernel objects, i.e. controller is passed into IKernel constructor and then passed all the way down through the constructors of KernelExecution, CellExecutionQueue, CellExecution etc. in order to create the notebook cell execution task
    - Controllers can only create execution tasks for the NotebookDocument that they are currently associated with in VS Code
    - Our code currently also starts a new kernel for each new NotebookDocument in `kernelProvider.getOrCreate`
    - Interactive window needs to be passed a controller ID at creation to pick up the right controller (alternative is `notebook.selectKernel` or similar after the window is created, but we've discussed this and decided not to do it). NotebookDocument is returned from `interactive.open`, so given the `jupyter-notebook` controller currently selected for a Jupyter notebook, we need to find the matching `interactive` controller that we told VS Code about, and make sure that the underlying kernel is reused
    - Note, this also means that if we want arbitrary notebooks to be able to start an interactive session, then all those extensions must ensure that they return the same instance for `jupyter-notebook` and `interactive` notebookTypes (not guaranteed to work OOB)

**Comment by ishipachev** [comment]
I would be glad to see this implemented.
My workflow around this feature would be like this:

- write some code in a notebook
- get some intermediate results and variables
- experiment with some variables or functions in console without creating additional code cell and cleaning it after
- put some outcomes of this experiments into notebook

Right now I can do it only in Interactive mode (with blocks starting with # %%), but this approach lacks of visual outputs I can have in notebooks.

**Comment by thehappycheese** [comment]
Yes please this is what I need! :)

**Comment by philffm** [comment]
Cheers! Seems like I managed to find a workaround. (can use the Jupyter Notebook variables in the interactive window now. A bit of a process but imho worth it :D)

## The process

1) Command + Shift + P
2) Jupyter: Create Interactive Window
3) Copy URL of Remote Jupyter Session
4) again: Command + Shift + P -> Jupyter: Specify local or remote Jupyter server for connections
5) Paste URL of Remote Jupyter Session
6) In the interactive window select the current session (remote)
7) That's it! The active variables from the Jupyter Notebook should now available in the interactive console

**Comment by Rchatru** [comment]
I have to try this, if it works it would be really wonderful. Thank you very much for sharing.

**Comment by notuntoward** [comment]
Yes, quick and easy like how Jupyterlab does its console.   Or Rstudio, or Matlab, or Spyder.

I'm trying out vscode -- and like it for features other than this -- but I'm really missing this fairly standard type of console.

**Comment by vttrifonov** [comment]
@philffm Thats cool! Just what I needed. 

Now if we can do the same dance but without having to start the Jupyter server separately... All within VSCode, please! I mean, sometimes it is definitely nice to have the remote server option but dealing with VSCode only will make life so easy.
